openshift-knative-operator/pkg/serving/extension.go

```diff
@@ -97,6 +97,15 @@ func (e *extension) Reconcile(ctx context.Context, comp v1alpha1.KComponent) err
 	// Use Kourier by default but allow a manual override.
 	common.ConfigureIfUnset(&ks.Spec.CommonSpec, "network", "ingress.class", "kourier.ingress.networking.knative.dev")
 
+	// Apply an Ingress config with Kourier enabled if nothing else is defined.
+	if ks.Spec.Ingress == nil {
```

Comment:
How was the behaviour before when no `.spec.ingress` was specified ? 

What happens if `.spec.ingress` is defined but neither `.spec.ingress.kourier` nor `.spec.ingress.istio` is given ?

Reply:
This is a little intricate so I'm going to try to answer extensively:

The `Spec.Ingress` API was introduced upstream fairly recently. We **don't** use it to install Kourier today. Instead, we're still installing Kourier through our own component (the knative-openshift component in this repo).

So, the `Spec.Ingress` API was a noop before this PR. Whatever you changed there, nothing would change, except if you set everything to false (which is equivalent to not mentioning the child keys, as they all default to false, since they are not pointers).

It's defaulting is defined here https://github.com/knative/operator/blob/39d7e473b24ae08101d440b43145d2d135a71fae/pkg/reconciler/knativeserving/ingress/ingress.go#L43-L59 and if you squint a little, you might spot a "bug": It says: If `Spec.Ingress` is not defined (note that it's a pointer for that reason), we assume you want to install Istio. Now, we don't want that. At first, I only specified an empty `IngressConfigurations` object. That causes everything to be `Enabled: false`, which should be fine because we don't actually install Kourier here, right? Wrong!

If everything is `Enabled: false`, this piece of code will return `mf.Any()` with an empty list of subfilters. `mf.Any` only returns true, if any of the predicates return true. Since we don't have any predicates, we're effectively going to filter away **everything** and won't install Knative Serving at all.

So that's a bug I'm going to go fix upstream and meanwhile we default to "Kourier" (in quotes because this doesn't actually install Kourier). That doesn't do any harm to the logic as it's going to filter the manifest for all resources that are **not** Kourier, which is exactly what we want in this case.

Reply:
I see and iiuc having both Kourier enabled here and in our own component doesn't collide as this config here is complementary and has no issues with our own way to install kourier.

Still my question would be what happens if the user provides an erroneous config like `.spec.ingress.courier.enabled=true`. I would expect this to be either ignore (so the same behaviour as if `.spec.ingress` was not given at all or an error raised.

However, I suspect from the code change above, providing this kind of config would prevent your workaround because `.spec.ingress` is != nil (but should be treated as it was nil).

Maybe a test for this kind of user error could be added, too, to validate any expected behaviour ?

Reply:
Currently, we'd indeed ignore it. I'll extend the behavior a little to catch the bug mentioned above for good measure.

Reply:
I don't quite want to surface an error via a webhook here because in theory, "all disabled" is a valid config I don't necessarily want to forbid (just maybe not officially support). So I'll extend the defaulting to Kourier to also be done (for now) if all ingresses are disabled.